File: lib/signalEngine.ts
import { ema, macd, rsi } from "./indicators";
import type { SignalRow, Signal, Strength } from "./types";

const toStrength = (score: number): Strength => {
  if (score >= 6) return "HIGH";
  if (score >= 4) return "MEDIUM";
  return "LOW";
};

const volatilityPct = (closes: number[]): number => {
  const tail = closes.slice(-24);
  if (tail.length < 3) return 0;
  const moves: number[] = [];
  for (let i = 1; i < tail.length; i += 1) {
    const prev = tail[i - 1];
    if (prev === 0) continue;
    moves.push(Math.abs(((tail[i] - prev) / prev) * 100));
  }
  return moves.length ? moves.reduce((a, b) => a + b, 0) / moves.length : 0;
};

const riskProfile = (signal: Signal, valueRsi: number, vol: number): { level: Strength; score: number; stop: string } => {
  let score = Math.round(25 + vol * 16);

  if (valueRsi >= 72 || valueRsi <= 28) score += 18;
  else if (valueRsi >= 67 || valueRsi <= 33) score += 10;
  if (signal === "HOLD") score += 8;

  score = Math.max(5, Math.min(95, score));

  const level: Strength = score >= 68 ? "HIGH" : score >= 38 ? "MEDIUM" : "LOW";
  if (signal === "HOLD") {
    return { level, score, stop: "Stop-loss acmaq ucun deqiq trade yoxdur" };
  }

  const base = Math.max(0.8, Math.min(4.8, vol * 1.7 + score / 95));
  const low = Math.max(0.6, Number((base * 0.8).toFixed(2)));
  const high = Math.min(6.8, Number((base * 1.25).toFixed(2)));
  return { level, score, stop: `${low}% - ${high}%` };
};

export function buildSignal(symbol: string, interval: string, closes: number[], timestampIso: string): SignalRow {
  const emaFastSeries = ema(closes, 20);
  const emaSlowSeries = ema(closes, 50);
  const sparkline = closes.slice(-20);

  if (!emaFastSeries.length || !emaSlowSeries.length) {
    return {
      symbol,
      signal: "HOLD",
      strength: "LOW",
      score: 0,
      reasons: ["Hec yeterli sayda candle yoxdur"],
      sparkline,
      risk_level: "LOW",
      risk_score: 10,
      stop_loss_range: "Stop-loss acmaq ucun deqiq trade yoxdur",
      snapshot: {
        close_price: closes.at(-1) ?? 0,
        rsi: 50,
        macd: 0,
        macd_signal: 0,
        timestamp: timestampIso,
      },
    };
  }

  const price = closes.at(-1) ?? 0;
  const emaFast = emaFastSeries.at(-1) ?? 0;
  const emaSlow = emaSlowSeries.at(-1) ?? 0;
  const valueRsi = rsi(closes);
  const { macd: valueMacd, signal: valueMacdSignal } = macd(closes);

  let buy = 0;
  let sell = 0;
  const reasons: string[] = [];

  if (price > emaFast && emaFast > emaSlow) {
    buy += 2;
    reasons.push("Qiymet EMA20 ve EMA50 ustundedir");
  } else if (price < emaFast && emaFast < emaSlow) {
    sell += 2;
    reasons.push("Qiymet EMA20 ve EMA50 altindadir");
  }

  if (valueMacd > valueMacdSignal) {
    buy += 2;
    reasons.push("MACD siqnal xettinin ustune kecib");
  } else if (valueMacd < valueMacdSignal) {
    sell += 2;
    reasons.push("MACD siqnal xettinin altina dusub");
  }

  if (valueRsi >= 45 && valueRsi <= 65) {
    buy += 1;
    reasons.push("RSI trend ucun normal zonadadir");
  }
  if (valueRsi >= 70) {
    sell += 1;
    reasons.push("RSI bazarin artiq alinmis oldugunu gosterir");
  }
  if (valueRsi <= 30) {
    buy += 1;
    reasons.push("RSI bazarin artiq satildigini ve rebound sansini gosterir");
  }

  let signal: Signal = "HOLD";
  let score = Math.max(buy, sell);

  if (buy >= sell + 2 && buy >= 4) {
    signal = "BUY";
    score = buy;
  } else if (sell >= buy + 2 && sell >= 4) {
    signal = "SELL";
    score = sell;
  }

  const strength = toStrength(score);
  const risk = riskProfile(signal, valueRsi, volatilityPct(closes));

  return {
    symbol,
    signal,
    strength,
    score,
    reasons: reasons.length ? reasons : ["Aydin istiqamet siqnali yoxdur"],
    sparkline,
    risk_level: risk.level,
    risk_score: risk.score,
    stop_loss_range: risk.stop,
    snapshot: {
      close_price: price,
      rsi: valueRsi,
      macd: valueMacd,
      macd_signal: valueMacdSignal,
      timestamp: timestampIso,
    },
  };
}

File: app/signal_engine.py
from app.indicators import ema, macd, rsi
from app.models import MarketSnapshot, TradeSignal


class SignalEngine:
    @staticmethod
    def _calc_volatility_pct(closes: list[float], window: int = 24) -> float:
        if len(closes) < 3:
            return 0.0

        tail = closes[-window:]
        moves: list[float] = []
        for prev, cur in zip(tail[:-1], tail[1:]):
            if prev == 0:
                continue
            moves.append(abs((cur - prev) / prev) * 100)

        if not moves:
            return 0.0
        return sum(moves) / len(moves)

    @staticmethod
    def _risk_profile(signal: str, latest_rsi: float, volatility_pct: float) -> tuple[str, int, str]:
        risk_score = int(25 + (volatility_pct * 16))

        if latest_rsi >= 72 or latest_rsi <= 28:
            risk_score += 18
        elif latest_rsi >= 67 or latest_rsi <= 33:
            risk_score += 10

        if signal == "HOLD":
            risk_score += 8

        risk_score = max(5, min(95, risk_score))

        if risk_score >= 68:
            risk_level = "HIGH"
        elif risk_score >= 38:
            risk_level = "MEDIUM"
        else:
            risk_level = "LOW"

        if signal not in {"BUY", "SELL"}:
            return risk_level, risk_score, "Stop-loss acmaq ucun deqiq trade yoxdur"

        base = max(0.8, min(4.8, (volatility_pct * 1.7) + (risk_score / 95)))
        low = max(0.6, round(base * 0.8, 2))
        high = min(6.8, round(base * 1.25, 2))
        stop_loss_range = f"{low}% - {high}%"
        return risk_level, risk_score, stop_loss_range

    def evaluate(self, symbol: str, interval: str, timestamp, closes: list[float]) -> TradeSignal:
        ema_fast_series = ema(closes, 20)
        ema_slow_series = ema(closes, 50)
        sparkline = closes[-20:]

        if not ema_fast_series or not ema_slow_series:
            snapshot = MarketSnapshot(
                symbol=symbol,
                interval=interval,
                timestamp=timestamp,
                close_price=closes[-1],
                ema_fast=0.0,
                ema_slow=0.0,
                rsi=50.0,
                macd=0.0,
                macd_signal=0.0,
            )
            return TradeSignal(
                symbol=symbol,
                signal="HOLD",
                strength="LOW",
                score=0,
                reasons=["Hec yeterli sayda candle yoxdur"],
                sparkline=sparkline,
                risk_level="LOW",
                risk_score=10,
                stop_loss_range="Stop-loss acmaq ucun deqiq trade yoxdur",
                snapshot=snapshot,
            )

        ema_fast = ema_fast_series[-1]
        ema_slow = ema_slow_series[-1]
        latest_rsi = rsi(closes)
        latest_macd, latest_macd_signal = macd(closes)
        price = closes[-1]

        snapshot = MarketSnapshot(
            symbol=symbol,
            interval=interval,
            timestamp=timestamp,
            close_price=price,
            ema_fast=ema_fast,
            ema_slow=ema_slow,
            rsi=latest_rsi,
            macd=latest_macd,
            macd_signal=latest_macd_signal,
        )

        buy_score = 0
        sell_score = 0
        reasons: list[str] = []

        if price > ema_fast > ema_slow:
            buy_score += 2
            reasons.append("Qiymet EMA20 ve EMA50 ustundedir")
        elif price < ema_fast < ema_slow:
            sell_score += 2
            reasons.append("Qiymet EMA20 ve EMA50 altindadir")

        if latest_macd > latest_macd_signal:
            buy_score += 2
            reasons.append("MACD siqnal xettinin ustune kecib")
        elif latest_macd < latest_macd_signal:
            sell_score += 2
            reasons.append("MACD siqnal xettinin altina dusub")

        if 45 <= latest_rsi <= 65:
            buy_score += 1
            reasons.append("RSI trend ucun normal zonadadir")
        if latest_rsi >= 70:
            sell_score += 1
            reasons.append("RSI bazarin artiq alinmis oldugunu gosterir")
        if latest_rsi <= 30:
            buy_score += 1
            reasons.append("RSI bazarin artiq satildigini ve rebound sansini gosterir")

        if buy_score >= sell_score + 2 and buy_score >= 4:
            signal = "BUY"
            score = buy_score
        elif sell_score >= buy_score + 2 and sell_score >= 4:
            signal = "SELL"
            score = sell_score
        else:
            signal = "HOLD"
            score = max(buy_score, sell_score)

        strength = "LOW"
        if score >= 6:
            strength = "HIGH"
        elif score >= 4:
            strength = "MEDIUM"

        volatility_pct = self._calc_volatility_pct(closes)
        risk_level, risk_score, stop_loss_range = self._risk_profile(signal, latest_rsi, volatility_pct)

        return TradeSignal(
            symbol=symbol,
            signal=signal,
            strength=strength,
            score=score,
            reasons=reasons or ["Aydin istiqamet siqnali yoxdur"],
            sparkline=sparkline,
            risk_level=risk_level,
            risk_score=risk_score,
            stop_loss_range=stop_loss_range,
            snapshot=snapshot,
        )
